go-controller/pkg/ovn/ha_master.go

```diff
@@ -0,0 +1,546 @@
+		kubeClient:      kubeClient,
+		ovnController:   ovnController,
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
+		isLeader:        false,
+		leaderElector:   nil,
+	}
+}
+
+// StartHAMasterController runs the replication controller
+func (hacontroller *HAMasterController) StartHAMasterController() error {
+	if hacontroller.manageDBServers {
+		// Always demote the OVN DBs to backup mode.
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		err := hacontroller.DemoteOVNDbs(invalidIPAddress, ovnNorthDbPort, ovnSouthDbPort)
+		if err != nil {
+			// If we are not able to communicate to the OVN ovsdb-servers,
+			// then it is better to return than continue.
+			// cmd/ovnkube.go will panic if this function returns error.
+			return err
+		}
+	}
+
+	// Set up leader election process first
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		haLeaderLockName,
+		hacontroller.kubeClient.CoreV1(),
+		nil,
+		resourcelock.ResourceLockConfig{
+			Identity:      hacontroller.nodeName,
+			EventRecorder: nil,
+		})
+
+	if err != nil {
+		return err
+	}
+
+	hacontrollerOnStoppedLeading := func() {
+		//This node was leader and it lost the election.
+		// Whenever the node transitions from leader to follower,
+		// we need to handle the transition properly like clearing
+		// the cache. It is better to exit for now.
+		// kube will restart and this will become a follower.
+		if hacontroller.manageDBServers {
+			// Stop ovn-northd and then exit.
+			_, _, _ = util.RunOVNNorthAppCtl("exit")
+		}
+		logrus.Infof("I (" + hacontroller.nodeName + ") am no longer a leader. Exiting")
+		os.Exit(1)
+	}
+
+	hacontrollerNewLeader := func(nodeName string) {
+		logrus.Infof(nodeName + " is the new leader")
+		wasLeader := hacontroller.isLeader
+
+		if hacontroller.nodeName == nodeName {
+			// Configure as leader.
+			logrus.Infof(" I (" + hacontroller.nodeName + ") won the election. In active mode")
+			err = hacontroller.ConfigureAsActive(nodeName)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				if hacontroller.manageDBServers {
+					// Stop ovn-northd before panicing.
+					_, _, _ = util.RunOVNNorthAppCtl("exit")
+				}
+				panic(err.Error())
+			}
+			hacontroller.isLeader = true
+		} else if wasLeader {
+			hacontrollerOnStoppedLeading()
+			// should not be reached
+			panic("This should not happen.")
+		} else {
+			// Configure as standby.
+			logrus.Infof(" I (" + hacontroller.nodeName + ") lost the election. In Standby mode")
+			ep, er := hacontroller.ovnController.kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+			if er == nil {
+				er = hacontroller.ConfigureAsStandby(ep)
+				if er != nil {
+					logrus.Errorf(er.Error())
+					if hacontroller.manageDBServers {
+						// Stop ovn-northd and then exit
+						_, _, _ = util.RunOVNNorthAppCtl("exit")
+					}
+					panic(er.Error())
+				}
+			}
+		}
```

Comment:
@numansiddique not sure when L137-149 is executed. Those lines are in the context of hacontrollerNewLeader() function. I am assuming this function will be called by the leader election framework on the leaders alone. If that is the case, then why would the leader every call `ConfigureAsStandby()`.

If I understand this PR correctly, the only place where we move the DB to standby() is in WatchOVNDbEndpoints() function.

Reply:
> @dcbw you stole my words :). That will be great.
> 
> Let us tackle `/var/run/ovn` in a separate PR and assume `/var/run/openvswitch` for now. Thanks.

Done. Updated the PR

Reply:
> @numansiddique not sure when L137-149 is executed. Those lines are in the context of hacontrollerNewLeader() function. I am assuming this function will be called by the leader election framework on the leaders alone. If that is the case, then why would the leader every call `ConfigureAsStandby()`.

Your assumption is wrong. hacontrollerNewLeader() will be called on all the ovnkube-master nodes. This is a notification to all the participants of the election that a new leader is chosen.

That's why you see `ConfigureAsStandby()` is called if the node lost the election.

It was suggested by @squeed to watch the endpoints as well. 

> 
> If I understand this PR correctly, the only place where we move the DB to standby() is in WatchOVNDbEndpoints() function.

DBs are moved to standby in 2 places - 
  1. When a leader changes and the hacontrollerNewLeader() is called and the node is not the leader.
  2. When the db endpoint update is detected. This could be because of (1)  and mostly redundant. But it's better to watch the endpoint also for any changes.